### ISSUE TITLE
Behavior of insert* commands for contenteditable=plaintext-only

### ISSUE THREAD

**Issue by zcorpan** [w3c/editing#419]
See https://bugzilla.mozilla.org/show_bug.cgi?id=1291467#c17 by @masayuki-nakano 

> I checked the behavior in Chrome roughly now, then, I see only the following commands work, so their definitions are required:
>
>   - Handling steps of `Document.execCommand("insertParagraph")` (Chrome/Safari handle this command differently from pressing Enter)
>   - Handling steps of `Document.execCommand("insertLineBreak")` (LF vs `<br>` (Chrome vs. Safari)
>   - Handling steps of the commands when caret is in special elements such as `<li>`
>   - Handling steps of `Document.execCommand("insertText")` (including line breaks causes insertParagraph in both, but I'm not sure it's intended since the behavior looks odd)
>   - Handling steps of `Document.execCommand("insertHTML")` (especially which elements need to be handled as line breaks)
>
> And the other commands (except `undo`, `redo` and clipboard commands) should be defined as NOOP (including not deleting `Selection`).

cc @howard-e

**Comment by zcorpan** [comment]
> Handling steps of `Document.execCommand("insertParagraph"`) (Chrome/Safari handle this command differently from pressing Enter)

https://software.hixie.ch/utilities/js/live-dom-viewer/saved/11814 (compare with [`contenteditable=true`](https://software.hixie.ch/utilities/js/live-dom-viewer/saved/11815)). Chrome and Safari insert a `div` and insert following sibling text to it (for both plaintext-only and true). Gecko (for `contenteditable=true`) results in two `div`s, one for `x` and one for `z`.

> Handling steps of `Document.execCommand("insertLineBreak"`) (LF vs `<br>` (Chrome vs. Safari)

LF seems more useful for plaintext-only.

**Comment by masayuki-nakano** [comment]
> > Handling steps of `Document.execCommand("insertLineBreak"`) (LF vs `<br>` (Chrome vs. Safari)
> 
> LF seems more useful for plaintext-only.

Yeah, if linebreaks are not preserved like `white-space: normal`, should be `<br>`? (I guess yes.)

And I wonder, if `contenteditable` value is changed from `""`/`"plaintext-only"` to the other value in a `beforeinput` event handler, how do Chrome/Safari handle it...? Especially when it's switched from `""` to `"plaintext-only"` and the command is not acceptable for the `plaintext-only`, e.g., `bold` command which is available with `Accel`+`B` except in Firefox.

**Comment by kojiishi** [comment]
Blink uses [`ShouldUseBreakElement`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/editing/commands/insert_line_break_command.cc;l=57?q=ShouldUseBreakElement&ss=chromium%2Fchromium%2Fsrc) to determine [whether to insert `<br>` or LF](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/editing/commands/insert_line_break_command.cc;l=96-99?q=ShouldUseBreakElement&ss=chromium%2Fchromium%2Fsrc); so it checks both `plaintextonly` and `white-space-collapse`.

**Comment by yisibl** [comment]
I've noticed that Chrome and Firefox behave differently with carriage returns(Enter key); Firefox inserts `<br>`, while Chrome does not.

Imagine if this feature is used in code highlighting scenarios, where users usually don't want the extra `<br>`.

![Image](https://github.com/user-attachments/assets/4cd6bbbc-f58f-4628-9252-1eae5e5a38e3)

This would seriously affect interoperability, and I think it's something that needs to be standardized in the specification as soon as possible.

**Comment by zcorpan** [comment]
@yisibl that seems to be because Chromium forces `white-space: pre-wrap` computed value for some specified values of `white-space`. WebKit and Gecko don't.

https://software.hixie.ch/utilities/js/live-dom-viewer/saved/13730

@kojiishi do you know why this happens in Chromium?

**Comment by kojiishi** [comment]
> I've noticed that Chrome and Firefox behave differently with carriage returns(Enter key); Firefox inserts <br>, while Chrome does not.

I haven't looked into the code, but at a glance, the Chrome behavior looks more natural to me because `<br>` isn't plaintext, no? Is the more detail of `plaintext-only` than the [contenteditable  content attribute](https://html.spec.whatwg.org/multipage/interaction.html#attr-contenteditable) defined somewhere?

/cc @tkent-google

**Comment by tkent-google** [comment]
Yeah, we assumed web developers didn't want `<br>`s in plaintext editing, and disallowed `white-space: normal` and `white-space: nowrap`.   If many web developers want `<br>`, it would be acceptable to change Chrome's behavior.

**Comment by zcorpan** [comment]
OK, though forcing computed style is "magic". Maybe it's enough to have a normal UA style rule?

```
[contenteditable="plaintext-only" i] { white-space: pre-wrap; }
```

If developers set it back to `normal` then they get `<br>`s.

@rniwa any opinions?

**Comment by masayuki-nakano** [comment]
The Chrome/Chromium's behavior violates the CSS rules and the style override hack seems odd if `white-space` style is partially specified in `contenteditable=plaintext-only` (explicitly setting `white-space` style is also overridden by Chrome/Chromium). Therefore, I filed [a chromium issue](https://issues.chromium.org/issues/413427424).

**Comment by masayuki-nakano** [comment]
Additionally, there is another issue of handling consecutive white-spaces. If `white-space` is `normal`, `nowrap` or `pre-line`, browsers need to normalize the white-spaces with pairs of an NBSP and an ASCII white-space to avoid collapsing them to a white-space.

**Comment by zcorpan** [comment]
> Maybe it's enough to have a normal UA style rule?

I submitted a PR for this: https://github.com/whatwg/html/pull/11351

**Comment by masayuki-nakano** [comment]
The white-space issue is now a real web-compat issue in TikTok's comment after inserting the mention target from their UI.
https://bugzilla.mozilla.org/show_bug.cgi?id=2015406#c17
